SupRsync

jlashner opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Agent for rsyncing files, checking md5sums, and removing from the local host.

## Description
<!--- Describe your changes in detail -->
The SupRsync agent and database are replacements for rsync cronjobs and pysmurf-archiver agent, and can be used to sync specific files to a remote host, compare checksums, and delete them from the localhost. 

This PR adds the SupRsync agent, an SqlAlchemy table / class to manage the files database, and changes to
the pysmurf-monitor which now interfaces with the suprsync table instead of a mysql database container.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Rsync cronjobs can't really be used reliably to checksum and delete archived files, so we have had lots of problems with servers filling up quickly with UFM data. This will solve that problem by automatically deleting files that have been successfully transfered after a specified amount of time.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
Added some tests to the agent's core functionality and am currently testing at UCSD on k2so

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [x] Unless I am preparing a release, I have opened this PR onto the `develop` branch.

jlashner commented:
Alright, made some changes after debuging more on the UCSD-K2SO system. Turns out that too many ssh cmds in a row will make a remote server block the connection, so I added more time in between each one, and gracefully handle timeouts.

I moved the file-handling logic to it's own helper class int the socs package so it can be easily tested and used by scripts other than the agent code.

I also added the `suprsync` script to trhe `bin` directory, which can be used to check / interact / debug the suprsync db. Currently it just has a few sub-commands for checking db info for specific files, getting the next file to be copied, and listing all files remaining to be copied.

BrianJKoopman commented on agents/suprsync/suprsync.py:
```diff
@@ -0,0 +1,152 @@
+            cmd_timeout=self.cmd_timeout, copy_timeout=self.copy_timeout
+        )
+
+        while self.running:
+            with srfm.Session.begin() as session:
+                file = srfm.get_next_file(
+                    self.archive_name, session=session,
+                    delete_after=self.delete_after,
+                    max_copy_attempts=self.max_copy_attempts)
+                if file is not None:
+                    try:
+                        handler.handle_file(file, session)
+                    except subprocess.TimeoutExpired:
+                        self.log.error(
+                            "Timed out when processing {path}",
+                            path=file.local_path)
+                        time.sleep(self.timeout_wait)
+                time.sleep(3)
```
Each iteration of this loop opens a session to the database every iteration, rsync's a single file, and takes 3-5+ seconds (3-5 seconds of sleep, plus operation time). This seems inefficient.

We touched up on this on the telecon today, but this should be reworked to handle more than one file at once. A good workflow I think would be to:
- Check the database for all files that need to be copied
- Assemble that list for passing to rsync
  - You could provide a long list of files at the command line, or write to a temporary file and use `--files-from`
  - The list will need to be built with paths relative to the remote-basedir, since that's the only thing specified on the other end, not the full path for each file like you're doing now.
  - This also avoids the whole need to create the remote location first for each subdirectory, as long as the remote-basedir exists (so you should still create that)
- rsync all files that need to be sent at once
- checksum remote files
  - I'm less sure of a way to do this efficiently remotely. Perhaps with the persistent ssh connection that Matthew suggested it won't be much of a problem
- update the database

You could run that loop once every few minutes or so.

BrianJKoopman reviewed (changes requested):
Thanks for the PR Jack! Most of my comments are made inline, perhaps the biggest one being about the loop structure within the Agent's run Process. Some other comments/questions:
- The monitor now uses a new database. Is there any special care that users of the monitor need to take when updating? i.e. should the old database be migrated, deleted, etc.
- What happens to files that exceed the max copy attempts? Do they need to be handled by hand?

jlashner commented on requirements.txt:
```diff
@@ -18,6 +18,9 @@ scipy
 pysnmp
 pysmi
 
+# SupRsync Agent
+sqlalchemy
+
```
I think only for docs, I can also probably just make this a mocked import

jlashner commented on socs/db/suprsync.py:
```diff
@@ -0,0 +1,355 @@
+        if file.local_md5sum == file.remote_md5sum:
+            if time.time() - file.timestamp > self.delete_after:
+                self.log.info(f"Deleting {file.local_path}")
+                os.remove(file.local_path)
```
I can't think of a way around this if the docker doesn't have permissions to operate on the file, except that if the agent doesn't have file permissions the delete_after param probably shouldn't be set. I can make a note of this in the docs though, unless you have a better idea of how this should be handled.

BrianJKoopman replied:
Right, there isn't much to do to solve it except fix the permissions. What we can do here though is catch the `PermissionError` and log that we couldn't delete the file so that a one off permission error doesn't crash the `run` Process.

jlashner replied:
Works for me

jlashner commented on agents/suprsync/suprsync.py:
```diff
@@ -0,0 +1,152 @@
+        )
+
+        while self.running:
+            with srfm.Session.begin() as session:
```
Yes all changes made within the context manager are committed if it exits successfully.

jlashner commented on docs/agents/suprsync.rst:
```diff
@@ -0,0 +1,184 @@
+           - "--site-hub=ws://${CB_HOST}:8001/ws"
+           - "--site-http=http://${CB_HOST}:8001/call"
+
+.. note::
+   If the SSH-key needed to access the remote server is not in the .ssh
+   directory, make sure it is being mounted into the docker-container
+   and that the ssh-key argument refers to its docker path.
```
Ok I will add this to the note. This can also easily be used for hk data if we want, but not sure where else it would be used.

jlashner commented on agents/pysmurf_monitor/pysmurf_monitor.py:
```diff
@@ -201,7 +182,9 @@ def make_parser(parser=None):
                         help="Specifies whether agent should create or update "
                              "pysmurf_files table if non exists.", default=True)
-
+    pgroup.add_argument('--db-path', type=str, default='/data/so/databases/suprsync.db',
+                        help="Path to suprsync sqlite database")
```
Where is the default user `ocs:ocs`? On the smurf-servers the `/data` directory is always owned by the cryo:smurf user. The monitor needs permission to create/modify the suprsync database, and the suprsync agent needs to be able to modify the database and also remove files on the host, so the easiest configuration is to make everyone the cryo:smurf user.

jlashner replied:
Ah I see, I changed the default block in the pysmurf-monitor to be cryo:smurf.

jlashner commented on socs/util.py:
```diff
@@ -24,6 +20,7 @@ def get_db_connection(**config):
     Same args as mysql.connector:
     https://dev.mysql.com/doc/connector-python/en/connector-python-connectargs.html
     """
+    import mysql.connector
```
This makes it so `mysql` isn't a requirement for anyone who wants to import the util module.

BrianJKoopman replied:
Mhm, I suppose, but `mysql-connector` is in the requirements file, so it should be installed with socs and the import shouldn't be a problem.

jlashner commented on docs/agents/suprsync.rst:
```diff
@@ -0,0 +1,184 @@
+
+.. note:: 
+
+   This agent does not remove empty directories, since I can't think of a
+   foolproof way to determine whether or not more files will be written to a
+   given directory, and pysmurf output directories will likely have
+   unregistered files that stick around even after copied files have been
+   removed. I think it's probably best to have a separate cronjob make that
+   determination and remove directory husks.
```
1. No, pysmurf only checks once when the instance is created, so if you delete the current output directory it will break things.
2. Yeah, I think just verifying that the directories are old enough or just that you're not deleting the most recent outputs directory for a given slot in case that's still in use.

BrianJKoopman replied:
Gotcha. How old is old enough to be deleted? I'm thinking of Katie's comments on needing certain files for the system to work, so they can't be moved off/deleted too soon.

jlashner commented on agents/suprsync/suprsync.py:
```diff
@@ -0,0 +1,152 @@
+                             "'<user>@<host>'). If None, will copy files locally")
+    pgroup.add_argument('--ssh-key', type=str,
+                        help="Path to ssh-key needed to access remote host")
+    pgroup.add_argument('--delete-after', type=float,
+                        help="Time (sec) after which this agent will delete "
+                             "local copies of successfully transfered files. "
+                             "If None, will not delete files.")
```
Fine with me

jlashner commented on socs/db/suprsync.py:
```diff
@@ -0,0 +1,355 @@
+                    continue
+            return f
+
+    def get_session(self):
+        """
+        Returns database session
+        """
+        return self.Session()
```
Nope! I'll just remove it.

jlashner commented on socs/db/suprsync.py:
```diff
@@ -0,0 +1,355 @@
+
+        session = self.Session()
+        session.add(file)
+        session.commit()
```
I have been using it a lot during local testing and it doesn't seem like it's slow or anything. The nice thing about not passing in a session is that you don't have to worry about committing afterwards, but I can add the option to pass in a session object.

jlashner commented on socs/db/suprsync.py:
```diff
@@ -0,0 +1,355 @@
+                "trying to copy", path=file.local_path)
+            file.local_md5sum = current_md5sum
+
+        if file.local_md5sum != file.remote_md5sum:
```
Yeah mostly it's because remote_md5sum is not set (and local_md5sum is required) but it always means the file needs to be copied or re-copied.

jlashner commented on socs/db/suprsync.py:
```diff
@@ -0,0 +1,355 @@
+        else:
+            dest = remote_path
+
+        cmd = ['rsync', '-t']
```
Seems like without "-pog" permissions are set to the user that you're copying to, which I believe is the behavior we want. (This changed a bit in the most recent commits)

jlashner commented on docs/agents/pysmurf/pysmurf-monitor.rst:
`````````````````````````````````````````diff
@@ -63,89 +63,26 @@ Where DB_HOST, DB, DB_USER, and DB_PW, SOCS_TAG, and CB_HOST are set in the
-````````````````````````
+Interfacing with the SupRsync Database
+````````````````````````````````````````
+The pysmurf-monitor agent now adds files
+to a specified suprsync database when a new pysmurf file is registered, or it
+receives a message from the publisher built into the smurf-streamer saying that
+a g3 file has been closed and finalized.
 
`````````````````````````````````````````
Good catch

BrianJKoopman replied:
Thanks. The text in the Docker Compose section around the configuration block still references the MYSQL related environment variables.

jlashner commented on socs/db/suprsync.py:
```diff
@@ -0,0 +1,355 @@
+            self.log.warn(
+                "Md5sum of {path} has changed! Resetting it in the db and "
+                "trying to copy", path=file.local_path)
+            file.local_md5sum = current_md5sum
```
Files are usually only registered after they finished writing. This mainly happens when a file is written and registered, and then written over or updated before being archived. This definitely isn't ideal, but has happened before and causes problems since if the local file changes before being archived, the agent will never be able to rsync successfully since it's comparing against a different checksum.

jlashner commented:
Thanks for all the great suggestions brian! I think I corrected / responded to all the comments, and changed the function to copy files in batches (or all at once) instead of one at a time.

I looked into it, and it seems like the --files-from option only lets you set the source to be copied, but doesn't allow you to set a custom destination for each file. In fact I couldn't find a good way with rsync to do this in a way that the path in the dest doesn't match the path of the source files. The solution I came up with was to generate a temporary directory and add a symlink to the remote_path from the base of the temp directory for each file. Then if I rsync the temp directory with the -L option (to follow symlinks and copy the actual files), we end up with all the files copied to the correct remote_paths on the remote server. I've tested this and it seems to work well, but I'm open to suggestions if you have a better solution. 

With these changes, copying a batch of files only uses a single rsync command and a single ssh command, so I think this addresses the problems I was seeing before.

BrianJKoopman commented on socs/db/suprsync.py:
```diff
@@ -0,0 +1,394 @@
+
+        Args
+        ----
+            archive_name (string):
+                Name of archive to get files from
+            session (sqlalchemy session):
+                Session to use to get files. If none is specified, one will
+                be created. You need to specify this if you wish to change
+                file data and commit afterwards.
+            max_copy_attempts (int):
+                Max number of failed copy atempts
+            num_files (int):
+                Number of files to return
```
Formatting here is slightly off. Arguments should be "name : type". See the docstring for `add_file()` above. The current syntax gets mangled when compiled by Sphinx.

jlashner replied:
ah this always confuses me. I swear I was just doing something where this was the correct format for docstrings but maybe in a different section or something...

BrianJKoopman replied:
This is *a* correct format. I think the issue might be the mix of formats. The rest of the file is in the Numpy style. This is the Google style.

I constantly reference this page (and the linked Google style at the top): https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html#example-numpy

BrianJKoopman commented on socs/db/suprsync.py:
```diff
@@ -0,0 +1,394 @@
+
+        Args
+        -----
+            archive_name (str):
+                Name of archive to pull files from
+            delete_after (float):
+                Time since creation (in seconds) for which it's ok to delete
+                files.
+            session (sqlalchemy session):
+                Session to use to query files.
```
Same docstring formatting issue as above.

BrianJKoopman reviewed (commented):
Sorry if this ended up split across two new reviews. I have some small lingering docstring comments. I think the only structural suggestion is catching the potential PermissionError when deleting files.

BrianJKoopman reviewed (approved):
Great, thanks for the last round of updates! It's looking good to me. I just merged in the develop branch, since there was a conflict in one file.

BrianJKoopman commented:
After discussion on the phone we decide we'll go ahead with the merge, but @jlashner will open an issues collecting updates that need to be made to this Agent to handle the database appropriately.
